File: examples/IntegratedNative/ios/Native/Components/RCA11yFocusWrapper/RCA11yFocusWrapper.m
//
//  RCA11yFocusWrapper.m
//  IntegratedNative
//
//  Created by Artur Kalach on 07.10.2022.
//

#import "RCA11yFocusWrapper.h"
#import <UIKit/UIKit.h>
#import <React/RCTViewManager.h>

@implementation RCA11yFocusWrapper

- (NSArray<id<UIFocusEnvironment>> *)preferredFocusEnvironments {
  if (self.myPreferredFocusedView == nil) {
      return @[];
    }
    return @[self.myPreferredFocusedView];
}
- (BOOL)canBecomeFocused {
    return self.canBeFocused;
}



- (void)didUpdateFocusInContext:(UIFocusUpdateContext *)context
       withAnimationCoordinator:(UIFocusAnimationCoordinator *)coordinator {
  if(!self.onFocusChange) {
    return;
  }
  
  if(context.nextFocusedView == self) {
    self.onFocusChange(@{ @"isFocused": @(YES) });
  } else if (context.previouslyFocusedView == self) {
    self.onFocusChange(@{ @"isFocused": @(NO) });
  }
}

- (void)didUpdateReactSubviews
{
  [super didUpdateReactSubviews];
  if (@available(iOS 14.0, *)) {
    self.focusGroupIdentifier =  [NSString stringWithFormat:@"app.group.%@", self.reactTag];
  }
}

@end

File: ios/Components/RCA11yFocusWrapperManager/RCA11yFocusWrapper.swift
//
//  RCTKeyboardFocusWrapper.swift
//  A11y
//
//  Created by Artur Kalach on 02.09.2022.
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation
import React
import UIKit

class RCA11yFocusWrapper: RCTView, KeyboardFocus {
  var myPreferredFocusedView: UIView! = nil
  
  @objc var onFocusChange: RCTBubblingEventBlock?
  @objc var canBeFocused: Bool = true
  
  override var preferredFocusEnvironments: [UIFocusEnvironment] {
    if (myPreferredFocusedView == nil) {
      return []
    }
    return [myPreferredFocusedView]
  }
  
  open override var canBecomeFocused: Bool {
    return canBeFocused
  }
  
  
  
  override func didUpdateReactSubviews() {
    super.didUpdateReactSubviews()

    if #available(iOS 14.0, *) {
      self.focusGroupIdentifier = "app.group.\(String(describing: self.reactTag))"
    }
  }
  
  override func didUpdateFocus(in context: UIFocusUpdateContext, with coordinator: UIFocusAnimationCoordinator) {
    
    if context.nextFocusedView == self {
      let params: [String: Bool] = ["isFocused": true]
      onFocusChange?(params)
    }
    else if context.previouslyFocusedView == self {
      let params: [String: Bool] = ["isFocused": false]
      onFocusChange?(params)
    }
  }
}
